python/pyiceberg/catalog/__init__.py

```diff
@@ -53,7 +54,15 @@ class Catalog(ABC):
 
     def __init__(self, name: str | None, **properties: str):
         self.name = name
-        self.properties = properties
+
+        if name is not None:
+            configuration = Config().get_catalog_config(name)
```

Comment:
I'd prefer not to load configuration from files every time a catalog is created. It also seems a bit strange to me to make this depend directly on config rather than passing it in.

On the other hand, I can see why it would be great to be able to rely on the environment config always being used so that callers don't need to instantiate their own `Config` and merge properties themselves.

How about using a factory method like this?

```python
_env_config = Config()

def load_catalog(name: str, **properties: str):
    env = _env_config.get_catalog_config(name)
    conf = merge_config(env or {}, properties)
    impl = conf['py-impl'] # not sure how we're doing this yet
    return impl(name, conf)
```

Reply:
Great suggestion, just implemented that. I left the `py-impl` out of scope; this PR is already getting far too big. 

While we're at it. One question, do we keep the `FileIO` at a catalog level? You could also specify it on a table level `create_table(..., properties={'py-impl': 'pyarrow'}` right? I was browsing the Java codebase, and it looks that's it is at a catalog level (which would make things easier).

Reply:
I took at look at the method, it's looking really good to me. Don't worry about `py-impl` -- I think the current approach of using the URI to decide is better. We can iterate from there.

For `FileIO`, you may need to change the IO based on the table you're using. A lot of people use the `HadoopFileIO` in Java, which supports many different storage schemes. But other use cases may need to choose between S3 and GCS per table. It's the responsibility of the catalog to pass a reasonable `FileIO` instance in when creating the table in Java. That's commonly the catalog's `FileIO`.

For Python, we can do things a little differently. I think it makes sense to combine information from the table and the environment to come up with the right `FileIO`. The table has storage and a policy for where to create new data files (`LocationProvider` in Java). I think that boils down to a storage scheme, like `s3`. The environment then determines what can be used to handle that scheme. In Python, pyarrow could be used, by there's also a S3 file system from fsspec.

In the short term, I think we should detect which file systems are available and use one that can handle the scheme for a table, where the scheme is taken from the table's location. Instantiating just the `PyArrowFileIO` works for now though.